airflow/providers/databricks/plugins/databricks_workflow.py

```diff
@@ -0,0 +1,453 @@
+    :return: None
+    """
+    hook = DatabricksHook(databricks_conn_id=databricks_conn_id)
+
+    repair_history_id = hook.get_latest_repair_id(databricks_run_id)
+    logger.debug("Latest repair ID is %s", repair_history_id)
+    logger.debug(
+        "Sending repair query for tasks %s on run %s",
+        tasks_to_repair,
+        databricks_run_id,
+    )
+
+    repair_json = {
+        "run_id": databricks_run_id,
+        "latest_repair_id": repair_history_id,
+        "rerun_tasks": tasks_to_repair,
+    }
+
+    return hook.repair_run(repair_json)
+
+
+def get_launch_task_id(task_group: TaskGroup) -> str:
+    """
+    Retrieve the launch task ID from the current task group or a parent task group, recursively.
+
+    :param task_group: Task Group to be inspected
+    :return: launch Task ID
+    """
+    try:
+        launch_task_id = task_group.get_child_by_label("launch").task_id  # type: ignore[attr-defined]
+    except KeyError as e:
+        if not task_group.parent_group:
+            raise AirflowException("No launch task can be found in the task group.") from e
+        launch_task_id = get_launch_task_id(task_group.parent_group)
+
+    return launch_task_id
+
+
+def _get_launch_task_key(current_task_key: TaskInstanceKey, task_id: str) -> TaskInstanceKey:
+    """
+    Return the task key for the launch task.
+
+    This allows us to gather databricks Metadata even if the current task has failed (since tasks only
+    create xcom values if they succeed).
+
+    :param current_task_key: The task key for the current task.
+    :param task_id: The task ID for the current task.
+    :return: The task key for the launch task.
+    """
+    if task_id:
+        return TaskInstanceKey(
+            dag_id=current_task_key.dag_id,
+            task_id=task_id,
+            run_id=current_task_key.run_id,
+            try_number=current_task_key.try_number,
+        )
+
+    return current_task_key
+
+
+@provide_session
+def get_task_instance(operator: BaseOperator, dttm, session: Session = NEW_SESSION) -> TaskInstance:
+    dag_id = operator.dag.dag_id
+    dag_run = DagRun.find(dag_id, execution_date=dttm)[0]
+    ti = (
+        session.query(TaskInstance)
+        .filter(
+            TaskInstance.dag_id == dag_id,
+            TaskInstance.run_id == dag_run.run_id,
+            TaskInstance.task_id == operator.task_id,
+        )
+        .one_or_none()
+    )
+    if not ti:
+        raise TaskInstanceNotFound("Task instance not found")
+    return ti
+
+
+def get_xcom_result(
+    ti_key: TaskInstanceKey,
+    key: str,
+    ti: TaskInstance | None,
+) -> Any:
+    result = XCom.get_value(
+        ti_key=ti_key,
+        key=key,
+    )
+    from airflow.providers.databricks.operators.databricks_workflow import WorkflowRunMetadata
+
+    return WorkflowRunMetadata(**result)
+
+
+class WorkflowJobRunLink(BaseOperatorLink, LoggingMixin):
+    """Constructs a link to monitor a Databricks Job Run."""
```

Comment:
Can you explain why we need all of this as plugin?
The goal is to allow extra link.
We have other providers that do that without plugins for example:
https://github.com/apache/airflow/tree/main/airflow/providers/google/cloud/links

Reply:
Along with the links, we also expose a RepairView here: https://github.com/apache/airflow/pull/40153/files#diff-dd75751ecf173b25fedb2c6c153a9bf55d44ccda0be11698e55920f52007e573R453 which sends repair requests based on the constructed URLs as part of the plugin utilities.

Reply:
I don't know databricks well enough so feel free to enlighten...
Why do we need to write so much more code and integration from our side to support extra links in databricks?
I am worried we are compensating for lack of features/capabilities from databricks and if so - this is not the path we should proceed with.

Reply:
It's not the lack of features in Databricks, but us trying to support the Databricks repair functionality from our operators that are run from within workflows

Reply:
I am conflicted about this one.
As you stated in the docs you wrote some deployments don't allow plugins.

Introducing plugins like so also couple it with fab which may be something we want to loose in Airflow 3. I am not sure this plugin should be part of the code base, maybe it should be served externally.
This may also relates to the actor issue I am raising in Airflow 3, dag authors wants this but deployment manger is not willing to support it.

I prefer to hear others take on this before we decide either way.

Reply:
Not a clear yes or no, but I just tried to search for plugins in airflow providers. Looks like only openlineage use it

Reply:
Yes. I think we should not do plugins that expose FAB Views. If anything. that calls for a new feature similar to extra link that could be built in Airflow. Not sure how, yet. because I do not understand what "repair" task is, but for me this looks like we are trying to do a very "poor" integration where we want to enable task clearing and re-running in external workflow, from within Airflow UI.  But maybe I do not understand the logic.

And it is very much something that I think we should be able to do in a generic way in Airflow (AKA workflows-in-workflow).  And i think implementing it as plugin rather than adding it as a core feture is a bad idea. 

Could you please explain a bit better what is the relation between Airflow tasks and DBWorkflow tasks in this case? Maybe explain scenario/use case - with explaining what happens during the execution and when users would like to use those links and what would happen when they do ?

Reply:
> where we want to enable task clearing and re-running in an external workflow from within the Airflow UI.

Yes, this is the right understanding and indeed what we are trying to do.

> @pankajkoti Could you please explain a bit better what is the relation between Airflow tasks and DBWorkflow tasks in this case? Maybe explain the scenario/use case, explaining what happens during execution, when users would like to use those links, and what would happen when they do?

Sure. [Workflows](https://docs.databricks.com/en/workflows/index.html) in Databricks consist of a group of tasks that may have dependencies between them. We represent these tasks as Airflow tasks. In Airflow, Databricks workflows are represented as TaskGroups (called DatabricksWorkflowTaskGroup). Within our Airflow DatabricksWorkflowTaskGroup, we have a launch task that runs these Databricks Workflow jobs using the [jobs endpoint](https://docs.databricks.com/en/workflows/jobs/create-run-jobs.html#run-a-job). In Databricks, a job can mean running a single task or a complex workflow containing multiple tasks. Here, we refer to the latter, a complex workflow consisting of multiple tasks. When triggered, the job runs the tasks in Databricks, and our Airflow tasks, which map 1:1 with the Databricks tasks, monitor their status and report success or failure upon completion.

Tasks in Databricks Workflows may fail for various reasons listed [here](https://docs.databricks.com/en/workflows/jobs/repair-job-failures.html#fix-the-cause-of-failure). The repair functionality in Databricks is analogous to our Airflow retry mechanism. We need to explore whether it's possible to integrate the repair functionality within our retry logic, though it might be challenging to add custom integration for making external calls during retries for specific operators. Additionally, Databricks provides a way to repair all failed tasks (except successful tasks) from a workflow with a single endpoint that accepts a list of tasks to re-run (repair). We also need to see how we can implement this from Airflow. Currently, the "repair all failed jobs links" functionality gets all failed tasks from our task group and calls the repair endpoint with those specific tasks.

Please let me know if anything is unclear or I missed explaining something in detail, and I can provide more details to help understand our intentions here.

Reply:
> Not a clear yes or no, but I just tried to search for plugins in airflow providers. Looks like only openlineage use it

just for records :), I believe the Hive provider also has a plugin with it.

Reply:
Hive plugin is legacy from Airflow 1.10

Reply:
Yeah. And the fact that no new plugin has been added  for 4 years says someething :).

Reply:
>Yes, this is the right understanding and indeed what we are trying to do.

Then why plugin? We can have operator for invoking the extral workflow and then you can just clear it. The button you try to add does something else... It invokes an external workflow regardless of the current Airflow task state. It may use its parameters but the exteral workflow is not attached to the Airflow task. The Airflow task can be Success while the external workflow is running.

Reply:
> Sure. [Workflows](https://docs.databricks.com/en/workflows/index.html) in Databricks consist of a group of tasks that may have dependencies between them....

So this is what I was thinking about. And this is what I see that should be "Workflow-in-workflow" -> having a task group that represents external workflow while tasks in Airflow represent 1:1 tasks in the external workflow is the way  how ALL external workflows should be mapped to Airlfow IMHO.

Now. The question is - can we introduce ExternalWorkflowTaskGroup (that DatabricksWorkflowTaskGroup would extend) and add the right features to it that the tasks inside it will be able to contribute interaction points in? I think it should be possible and quite "generic" so that other external workflows could follow the same patterns - without having to resort to plugins?

Reply:
I think it would be quite possible and relatively easy to turn it into a feature of Airlfow core.

Reply:
>  The Airflow task can be Success while the external workflow is running.

@eladkal this will not be the case as the Airflow task will only be marked as success when the corresponding task is success in the external workflow. The very responsibility of the Airflow task is to monitor the status of the external workflow task.

Reply:
Then i am missing something.
If the repair button can invoke job then you can do it at will. Is that right?

Reply:
yes, that is true and we're doing it so here https://github.com/apache/airflow/pull/40153/files#diff-dd75751ecf173b25fedb2c6c153a9bf55d44ccda0be11698e55920f52007e573R396

Reply:
So you can have Airflow task in Success state then come back to it few days later and use this button to invoke remote job without rerunning the Airflow task.
I don't think that is ideal use case for the extra link button.

We should come up with better support for such feature that does not involve plugins

Reply:
hi, I have attempted to demo the Databricks Workflow represented within Airflow and repair functionality in a 7.5 minutes video attached below. I hope it could help further in what I tried to say previously in the [comment](https://github.com/apache/airflow/pull/40153#discussion_r1660225043).


https://github.com/apache/airflow/assets/10206082/8ecaa46b-d335-4037-b880-ba16674914da

Hope you get a chance to go through the video and we get to discuss this further for finding a resolution.

Reply:
Thanks a lot for the video @pankajkoti . It's been super helpful.

Yes. This is working exactly as I'd imagined it, so it confirms my assesment - that this is precisely something we should build into airflow core.

However,

I do see that as an upcoming Airlfow 3.1+ feature, not something that we will be able to im[plement now, nor even as part of 3.0 effort. Adding all the interaction possibilities - launching the job behind the scenes, observing and monitoring individual tasks run in the external workflow, ability to clear and "re-run" (i.e. repair) individual tasks are all the features I'd imagined "workflow-in-workfow" case should have as a core feature and well designed API. But - we have other things as priority to implement for Airllow 3, and also such feature will only be available in Airflow 3.x (when it is implement and if we agree that such functionality is something we should implement in Airlfow).

Implementing it as plugin in the way you described is more of a band-aid and it's not easily replicable to other workflows without similar complexity.

But it has an undeniable advantage that we can make it works now in Airflow 2, put it in the hands of our users today and learn from it when we attempt to implement the Airflow 3 "generic" solution. Which we might not be doing for a year or so, so waiting for it does not allow us to learn from the ways our users would like to use it. 

So - overall - I'd be for doing it this way with few caveats:

* we commit to make it Airflow 3 -plugin compliant (i.e. future version of the databricks workflow provider will have the airflow 3 - compatible plugin variant). I uderstand this is "given"
* we plan to have a workstream (Airflow 3.1+) where we turn such approach (also taking patterns from Cosmos integration) where we turn it into a generic solution

We do not have to do the "generic" attempts now - just plan it as a natural evolution of this approach and whenever we get to Airflow 3.1+ to implement it (or generally where we fill we would like to do simlar approach for another workflow) - we should THEN sit down and decompose it into what would it mean for Core Airflow to support this kind of "external workflow" integration in Airflow.

That would by my ask at this moment. I think the way it is implemented now - where we use plugins and make such workflow available for Airflow 2 users is also a nice "teaser"  - it would give us a feedback from the users on how useful and needed it is, so that we could have more clarity on what exactly is needed, what works, what kind of problems we need to solve when we attempt to make it Airflow 3 "feature".

Reply:
Thank you @potiuk for your insightful comment. We(at Astronomer) truly appreciate your feedback and are glad that the video was helpful.

We agree with your assessment and are committed to complying with the Airflow 3 approach. We understand the importance of having a well-designed core feature for the "workflow-in-workflow" case in future versions of Airflow.

While we recognize that implementing this as a plugin in Airflow 2 is more of a temporary solution, it provides immediate value and allows us to gather valuable user feedback. This will be instrumental when we develop the Airflow 3.x generic solution.

We are fully committed to migrating the plugin to use the Airflow core feature once it is ready. Additionally, we will ensure that the future versions of the Databricks workflow provider will have an Airflow 3-compatible plugin variant.
We appreciate your guidance and will help contributing in the plan for creating workstream for Airflow 3.1+ to evolve this approach into a generic solution. Thanks again for your input.

Reply:
@eladkal when you have a chance and find some time, would you please be able to check the demo in [comment](https://github.com/apache/airflow/pull/40153#discussion_r1662876620) which explain how the workflow works and the integrated functionalities?

Looking forward to hearing your thoughts with the current approach and future plan being discussed here based on Jarek's suggestion.

Reply:
Thanks to @tatiana, we also have logged a ticket https://github.com/apache/airflow/issues/40587 for the future work to reimplement this.

Reply:
I also added a workstream in Airlfow 3.1+ that explains how inspiration from this PR and Astronomer's Cosmos could be taken to next level in the future version of Airflow: https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+3+Workstreams#Airflow3Workstreams-Othercandidates.1